### PR TITLE
Refactor CollectionVersion Upload to use pulpcore machinery 

### PR DESCRIPTION
**Pull request by gerrod3**
fixes: #1175

### CODE REVIEW

File: pulp_ansible/app/galaxy/views.py
Comment (mdellweg) [inline pulp_ansible/app/galaxy/views.py:189]: I think, this belongs in the serializer.validate.

File: pulp_ansible/app/serializers.py
Comment (mdellweg) [inline pulp_ansible/app/serializers.py:386]: :+1:

File: pulp_ansible/tests/functional/api/collection/v3/test_collection.py
Comment (mdellweg) [inline pulp_ansible/tests/functional/api/collection/v3/test_collection.py:312]: what kind of a fixture is "artifact"? If it's a pulp_ansible one can we name it `collection_artifact`?

File: pulp_ansible/tests/functional/api/collection/v3/test_collection.py
Comment (mdellweg) [inline pulp_ansible/tests/functional/api/collection/v3/test_collection.py:201]: If these do not assert anymore, i think we are seeing a change behavioural change unrelated to the bug / improvement advertised in the PR.

File: pulp_ansible/tests/functional/api/collection/v3/test_collection.py
Comment (mdellweg) [reply pulp_ansible/tests/functional/api/collection/v3/test_collection.py:312]: Are the changes in this file supposed to be part of this PR at all?

Comment (gerrod3) [reply pulp_ansible/tests/functional/api/collection/v3/test_collection.py:312]: Yes, I realized the tests for collection upload were wrong and incorrectly passing when it shouldn't.

File: pulp_ansible/tests/functional/api/collection/v3/test_collection.py
Comment (gerrod3) [reply pulp_ansible/tests/functional/api/collection/v3/test_collection.py:201]: The asserts were never being ran in the first place. `collection_upload` is a `CollectionImport` response and does not have the fields in `COLLECTION_METADATA` : `namespace, name, version`. So commenting them out does nothing I could have left it as is.

Comment (mdellweg) [reply pulp_ansible/tests/functional/api/collection/v3/test_collection.py:202]: Are you saying the keys should have been there? Then we should not guard with the "if". If not can we just remove the block completely?

Comment (gerrod3) [reply pulp_ansible/tests/functional/api/collection/v3/test_collection.py:202]: No idea if they should have been there. But they will be once I add `Namespace`s.

File: pulp_ansible/app/serializers.py
Comment (mdellweg) [inline pulp_ansible/app/serializers.py:489]: Does it make sense to add the add-to-repository-in-a-subtask feature to pulpcore?

File: pulp_ansible/app/tasks/upload.py
Comment (mdellweg) [inline pulp_ansible/app/tasks/upload.py:17]: This is the thing we call in `deferred_validate` now? Great achievement!

File: pulp_ansible/app/serializers.py
Comment (mdellweg) [inline pulp_ansible/app/serializers.py:427]: Shouldn't we still call `super` first?

File: pulp_ansible/app/galaxy/v3/views.py
Comment (gerrod3) [reply pulp_ansible/app/galaxy/v3/views.py:473]: So I can hide the Pulp-specific fields: `artifact`, `upload`, and `repository` from the generated schema. I tried to just not include them in the `Meta` class, but then it would error when `artifact` or `repository` was set by the viewset's logic.

File: pulp_ansible/app/serializers.py
Comment (gerrod3) [reply pulp_ansible/app/serializers.py:489]: I'm thinking about it. I already know how I am going to implement the feature in pulp_ansible, but putting it into pulpcore would make it much more simple.

File: pulp_ansible/app/serializers.py
Comment (gerrod3) [reply pulp_ansible/app/serializers.py:427]: I'm thinking of removing this check and instead over-riding the `update` method to raise an error if called. `validate` should only be called on `create` or `update` so if `self.instance` is set then `update` will be called and we shouldn't allow it.

Comment (mdellweg) [reply pulp_ansible/app/serializers.py:427]: Hmm. Hard question. From the API we should never allow updates. But we want to allow for a repair command. And it would be super neat if that repair command could just use this serializer to actually _call_ the `process_collection` part again. No?

Comment (gerrod3) [reply pulp_ansible/app/serializers.py:427]: Yes that would be a good use. There would still need to be other fields that would need to be set if you wanted to call `super` in that case (`artifact`, `expected_namespace/name/version/`) else the serializer would fail validation. I'll remove the check and put a TODO comment to revisit this for a future issue (#1106).

**Review by gerrod3 — COMMENTED** [review]
@mdellweg @newswangerd Ready for review.

Summary of changes:

- `/ansible/collections/` & Galaxy V2 Collection upload have been deprecated.
- `CollectionVersionUploadSerializer` now has all the old fields of `CollectionOneShotSerializer` and the entirety of the collection upload logic needed.
- Galaxy V3 Collection upload is now a `SingleArtifactContentUploadViewSet` and uses the `CollectionVersionUploadSerializer` to perform the collection upload.
- `CollectionVersionSerializer` now has `CollectionVersionUploadSerializer` as a base class so the Pulp CV api uses the same base upload logic as the Galaxy V3 api.
- Pulp CV api now accepts `artifact` and `upload` fields when uploading a collection.

File: pulp_ansible/app/galaxy/mixins.py
Comment (mdellweg) [reply pulp_ansible/app/galaxy/mixins.py:25]: I think this is by design. As the task will call the same serializer again (including the deferred validate steps) so it needs the raw data.

File: pulp_ansible/app/galaxy/serializers.py
Comment (mdellweg) [inline pulp_ansible/app/galaxy/serializers.py:197]: Is it safe to just take the existing artifact at this time? (The user uploaded the proper file, right?)

File: pulp_ansible/app/galaxy/v3/views.py
Comment (mdellweg) [reply pulp_ansible/app/galaxy/v3/views.py:473]: It think, this is dangerous. Just not adding them to the docs, means that the serializer still accepts them. And it shouldn't!?

File: pulp_ansible/app/serializers.py
Comment (mdellweg) [inline pulp_ansible/app/serializers.py:428]: ... "in task context"; "not in request context"

File: pulp_ansible/app/serializers.py
Comment (mdellweg) [inline pulp_ansible/app/serializers.py:432]: You are basically inverting the "deferred_validate" logic here. Can we just start with the raw data again in the task, so we can be sure that we validate everything in one go?

File: pulp_ansible/app/serializers.py
Comment (mdellweg) [inline pulp_ansible/app/serializers.py:453]:
```suggestion
            if CollectionVersion.objects.filter(**{f: data[f"expected_{f}"] for f in fields}).exists():
```

File: pulp_ansible/app/serializers.py
Comment (mdellweg) [reply pulp_ansible/app/serializers.py:454]: The existing one? Only if we wanted to use it, IMHO.

File: pulp_ansible/app/serializers.py
Comment (gerrod3) [reply pulp_ansible/app/serializers.py:432]: Yes, I can rewrite it. Just need to take into account not having the artifact filename on the second validate.

File: pulp_ansible/app/galaxy/v3/views.py
Comment (gerrod3) [reply pulp_ansible/app/galaxy/v3/views.py:473]: You are right. I'll add validation for this serializer.

File: pulp_ansible/app/galaxy/serializers.py
Comment (mdellweg) [reply pulp_ansible/app/galaxy/serializers.py:197]: Do we plan to (eventually?) change that task then too?

Comment (gerrod3) [reply pulp_ansible/app/galaxy/serializers.py:197]: I eventually plan to remove the task as it is duplicate code now and eventually remove the deprecated endpoints that call it. Git sync still uses it as well so that will also require a separate PR to refactor it.

File: pulp_ansible/app/serializers.py
Comment (mdellweg) [inline pulp_ansible/app/serializers.py:389]: Isn't this where we also call the import_collection to extract all the metadata? That process that may be offloaded into a different application cluster?

File: pulp_ansible/app/serializers.py
Comment (mdellweg) [reply pulp_ansible/app/serializers.py:613]: I'm still not sure we need this. Is it there to support code we plan to deprecate/delete? If i read correctly it allows the serializer to consume validated data, right? I do not think they are designed that way.
My thinking is if we hand the same raw request data that we get in the viewset to the task, then the serializer in the task can validate in the same way plus the additional deferred_validate step. No need for any serializer field to behave differently.
Not blocking on this now. But can you confirm that this will fall for another refactor to come?

**Review by mdellweg — APPROVED** [review]
I'm still a bit confused by the big number of serializers we have for creating CollectionVersions. Maybe we need individual ones for each version of the api, and that is fine.
We still need confirmation from the galaxy side before we can merge this.

File: pulp_ansible/app/serializers.py
Comment (gerrod3) [reply pulp_ansible/app/serializers.py:389]: Yes, I wrote these steps describing how the pulpcore machinery works since I feel it is not clear from just looking at the serializer what is going on.

File: pulp_ansible/app/serializers.py
Comment (gerrod3) [reply pulp_ansible/app/serializers.py:613]: Yes it most likely will be removed once the `import_collection` code is finally removed. `read_only` fields are not validated when initialized in a serializer and they don't appear in the `validate` method.

**Review by mdellweg — APPROVED** [review]

**Pull request merged**
